File: benchmarks/realworld/compress/C/compress.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>

#define BUF_SIZE 8192

double now_sec() {
    struct timespec ts;
    clock_gettime(CLOCK_MONOTONIC, &ts);
    return ts.tv_sec + ts.tv_nsec / 1e9;
}

// 简化 RLE 压缩（遇到重复字节则编码为 [count][byte]）
size_t simple_compress(const unsigned char *in, size_t in_len, unsigned char *out) {
    size_t out_idx = 0;
    for (size_t i = 0; i < in_len;) {
        unsigned char b = in[i];
        size_t count = 1;
        while (i + count < in_len && in[i + count] == b && count < 255) count++;
        out[out_idx++] = count;
        out[out_idx++] = b;
        i += count;
    }
    return out_idx;
}

int main(int argc, char *argv[]) {
    if (argc != 2) {
        printf("Usage: %s <file>\n", argv[0]);
        return 1;
    }

    FILE *fp = fopen(argv[1], "rb");
    if (!fp) {
        perror("fopen");
        return 1;
    }

    fseek(fp, 0, SEEK_END);
    size_t fsize = ftell(fp);
    fseek(fp, 0, SEEK_SET);

    unsigned char *input = malloc(fsize);
    unsigned char *output = malloc(fsize * 2); // worst case

    fread(input, 1, fsize, fp);
    fclose(fp);

    double t_start = now_sec();
    size_t out_size = simple_compress(input, fsize, output);
    double t_end = now_sec();

    printf("Compression time: %.6f sec\n", t_end - t_start);
    printf("Original size: %zu bytes\nCompressed size: %zu bytes\n", fsize, out_size);

    free(input);
    free(output);
    return 0;
}

File: benchmarks/realworld/compress/Zig/compress.zig
const std = @import("std");

fn nowSec() f64 {
    return @as(f64, @floatFromInt(std.time.microTimestamp())) / 1_000_000.0;
}

fn simpleCompress(allocator: std.mem.Allocator, input: []const u8) ![]u8 {
    var output = std.ArrayList(u8).init(allocator);
    defer output.deinit();
    
    var i: usize = 0;
    while (i < input.len) {
        const b = input[i];
        var count: usize = 1;
        while (i + count < input.len and input[i + count] == b and count < 255) {
            count += 1;
        }
        try output.append(@intCast(count));
        try output.append(b);
        i += count;
    }
    return output.toOwnedSlice();
}

pub fn main() !void {
    const stdout = std.io.getStdOut().writer();
    
    var arena = std.heap.ArenaAllocator.init(std.heap.page_allocator);
    defer arena.deinit();
    const allocator = arena.allocator();

    const args = try std.process.argsAlloc(allocator);
    defer std.process.argsFree(allocator, args);

    if (args.len != 2) {
        try stdout.print("Usage: {s} <file>\n", .{args[0]});
        std.process.exit(1);
    }

    const input_data = try std.fs.cwd().readFileAlloc(allocator, args[1], std.math.maxInt(usize));
    
    const t_start = nowSec();
    const output_data = try simpleCompress(allocator, input_data);
    const t_end = nowSec();

    try stdout.print("Compression time: {d:.6} sec\n", .{t_end - t_start});
    try stdout.print("Original size: {d} bytes\n", .{input_data.len});
    try stdout.print("Compressed size: {d} bytes\n", .{output_data.len});
}
